src/opensles/AudioOutputStreamOpenSLES.cpp

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright 2017 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cassert>
+
+#include <SLES/OpenSLES.h>
+#include <SLES/OpenSLES_Android.h>
+
+#include "oboe/AudioStreamBuilder.h"
+#include "AudioOutputStreamOpenSLES.h"
+#include "AudioStreamOpenSLES.h"
+#include "OpenSLESUtilities.h"
+#include "OutputMixerOpenSLES.h"
+
+using namespace oboe;
+
+AudioOutputStreamOpenSLES::AudioOutputStreamOpenSLES(const AudioStreamBuilder &builder)
+        : AudioStreamOpenSLES(builder) {
+}
+
+AudioOutputStreamOpenSLES::~AudioOutputStreamOpenSLES() {
+}
+
+// These will wind up in <SLES/OpenSLES_Android.h>
+#define SL_ANDROID_SPEAKER_QUAD (SL_SPEAKER_FRONT_LEFT | SL_SPEAKER_FRONT_RIGHT \
+ | SL_SPEAKER_BACK_LEFT | SL_SPEAKER_BACK_RIGHT)
+
+#define SL_ANDROID_SPEAKER_5DOT1 (SL_SPEAKER_FRONT_LEFT | SL_SPEAKER_FRONT_RIGHT \
+ | SL_SPEAKER_FRONT_CENTER  | SL_SPEAKER_LOW_FREQUENCY| SL_SPEAKER_BACK_LEFT \
+ | SL_SPEAKER_BACK_RIGHT)
+
+#define SL_ANDROID_SPEAKER_7DOT1 (SL_ANDROID_SPEAKER_5DOT1 | SL_SPEAKER_SIDE_LEFT \
+ |SL_SPEAKER_SIDE_RIGHT)
+
+int AudioOutputStreamOpenSLES::chanCountToChanMask(int chanCount) {
+    int channelMask = 0;
+
+    switch (chanCount) {
+        case  1:
+            channelMask = SL_SPEAKER_FRONT_CENTER;
+            break;
+
+        case  2:
+            channelMask = SL_SPEAKER_FRONT_LEFT | SL_SPEAKER_FRONT_RIGHT;
+            break;
+
+        case  4: // Quad
+            channelMask = SL_ANDROID_SPEAKER_QUAD;
+            break;
+
+        case  6: // 5.1
+            channelMask = SL_ANDROID_SPEAKER_5DOT1;
+            break;
+
+        case  8: // 7.1
+            channelMask = SL_ANDROID_SPEAKER_7DOT1;
+            break;
+    }
+    return channelMask;
+}
+
+Result AudioOutputStreamOpenSLES::open() {
+    Result oboeResult = AudioStreamOpenSLES::open();
+    if (Result::OK != oboeResult)  return oboeResult;
+
+    SLresult result = OutputMixerOpenSL::getInstance()->open();
+    if (SL_RESULT_SUCCESS != result) {
+        AudioStreamOpenSLES::close();
+        return Result::ErrorInternal;
+    }
+
+    SLuint32 bitsPerSample = getBytesPerSample() * OBOE_BITS_PER_BYTE;
```

Comment:
Use `constexpr` for this oboe constant and add it to the oboe namespace. The usual convention is to prefix constants with k so this would end up as `oboe::kBitsPerByte`

Reply:
Done